mypy/newsemanal/semanal.py

```diff
@@ -3731,9 +3752,25 @@ def lookup(self, name: str, ctx: Context,
         if self.type and not self.is_func_scope() and name in self.type.names:
             node = self.type.names[name]
             if not node.implicit:
-                return node
-            implicit_name = True
-            implicit_node = node
+                # Only allow access to class attributes textually after
+                # the definition, so that it's possible to fall back to the
+                # outer scope. Example:
+                #
+                #     class X: ...
+                #
+                #     class C:
+                #         X = X  # Initializer refers to outer scope
+                #
+                # Nested classes are an exception, since we want to support
+                # arbitrary forward references in type annotations.
+                if (node.node is None
+                        or node.node.line < self.statement.line
+                        or isinstance(node.node, TypeInfo)):
```

Comment:
Node may be also `TypeAlias`, for example:
```python
class C:
    x: A
    A = List[int]
```
In general this looks a bit ad-hoc. Would it be better instead to add a flag to lookup function that indicates whether lookup happens from type context or from runtime context? Or will this cause problems in situations like this:
```python
class A: ...
class C:
    A = List[A]  # A looked-up in type context on r.h.s.
```